README.md
redmin_demo
===========

Project Management System

doc/CHANGELOG
== Redmine changelog

Redmine - project management software
Copyright (C) 2006-2012  Jean-Philippe Lang
http://www.redmine.org/

== 2012-11-17 v2.1.3

* Defect #12050: :export links to repository files lead to a 404 error
* Defect #12189: Missing tmp/pdf directory
* Defect #12195: Javascript error with IE7 / IE8 on new issue form
* Defect #12196: "Page not found" on OK button in SCM "View all revisions" page
* Defect #12199: Confirmation message displayed when clicking a disabled delete link in the context menu
* Defect #12231: Hardcoded "Back" in Repository
* Defect #12294: Incorrect german translation for "registered" users filter
* Defect #12349: Watchers auto-complete search on non-latin chars
* Defect #12358: 'None' grouped issue list section should be translated
* Defect #12359: Version date field regex validation accepts invalid date
* Defect #12375: Receiving mail subject encoding broken (MOJIBAKE) in some cases on Ruby 1.8
* Patch #9732: German translations
* Patch #12021: Russian locale translations
* Patch #12188: Simplified Chinese translation with zh.yml file based on Rev:10681
* Patch #12235: German translation for 2.1-stable
* Patch #12237: Added German Translation

== 2012-09-30 v2.1.2

* Defect #11929: XSS vulnerability in Redmine 2.1.x

== 2012-09-30 v2.1.1

* Defect #11290: ParseDate missing in Ruby 1.9x
* Defect #11844: "load_default_data" rake task fails to print the error message if one occurs
* Defect #11850: Can't create a user from ldap by on-the-fly on the redmine server using URI prefix
* Defect #11872: Private issue visible to anonymous users after its author is deleted
* Defect #11885: Filter misses Selectionfield on IE8
* Defect #11893: New relation form Cancel link is broken with Chrome 21
* Defect #11905: Potential "can't dup NilClass" error in UserPreference
* Defect #11909: Autocomplete results not reset after clearing search field
* Defect #11922: bs.yml and de.yml lead to error by number_to_currency()
* Defect #11945: rake task prints "can't convert Errno::EACCES into String" in case of no permission of public/plugin_assets
* Defect #11975: Undefined status transitions allowed in workflow (author of issue changes when selecting a new status)
* Defect #11982: SCM diff view generates extra parameter for switching mode
* Patch #11897: Traditional Chinese language file (to r10433)

== 2012-09-16 v2.1.0

* Defect #2071: Reordering priority-enumerations breaks alternate-theme's issue-colouring
* Defect #2190: Month names not translated to german
* Defect #8978: LDAP timeout if an LDAP auth provider is unreachable
* Defect #9839: Gantt abbr of weekday should not be necessarily the first letter of the long day name
* Defect #10928: Documentation about generating a plugin is not up-to-date
* Defect #11034: TLS configuration documentation for Rails 3
* Defect #11073: UserCustomField order_statement returns wrong output
* Defect #11153: Default sorting for target version is DESC instead of ASC
* Defect #11207: Issues associated with a locked version are not copied when copying a project
* Defect #11304: Issue-class: status-1, status-2 etc. refer to status position instead of status id
* Defect #11331: Openid registration form should not require user to enter password
* Defect #11345: Context menu should show shared versions when editing issues from different projects
* Defect #11355: Plain text notification emails content is HTML escaped
* Defect #11388: Updating a version through rest API returns invalid JSON
* Defect #11389: Warning in awesome_nested_set.rb
* Defect #11503: Accessing /projects/:project/wiki/something.png fails with error 500
* Defect #11506: Versions that are not shared should not be assignable when selecting another project
* Defect #11508: Projects not ordered alphabetically after renaming project
* Defect #11540: Roadmap anchor links can be ambigous
* Defect #11545: Overwriting existing method Issue.open
* Defect #11552: MailHandler does not match assignee name with spaces
* Defect #11571: Custom fields of type version not proper handled in receiving e-mails
* Defect #11577: Can't use non-latin anchor in wiki
* Defect #11612: Revision graph sometimes broken due to raphael.js error
* Defect #11621: Redmine MIME Detection Of Javascript Files Non-Standard
* Defect #11633: Macro arguments should not be parsed by text formatters
* Defect #11662: Invalid query returned from Issues.visible scope after accessing User#projects_by_role with a role that is not present
* Defect #11691: 404 response when deleting a user from the edit page
* Defect #11723: redmine:send_reminders notification misses if assignee is a group
* Defect #11738: Batch update of issues clears project path
* Defect #11749: Redmine.pm: HEAD is not considered as a read-only method
* Defect #11814: Date picker does not respect week start setting
* Feature #703: Configurable required fields per tracker/status/role
* Feature #1006: Display thumbnails of attached images
* Feature #1091: Disabling default ticket fields per tracker
* Feature #1360: Permission for adding an issue to a version.
* Feature #3061: Let macros optionally match over multiple lines and ignore single curly braces
* Feature #3510: Inserting image thumbnails inside the wiki
* Feature #3521: Permissions for roles to change fields per tracker/status
* Feature #3640: Freeze / Close Projects
* Feature #3831: Support for subforums
* Feature #6597: Configurable session lifetime and timeout
* Feature #6965: Option to Copy Subtasks when copying an issue
* Feature #8161: Ability to filter issues on project custom fields
* Feature #8577: "Private" column and filter on the issue list
* Feature #8981: REST Api for Groups
* Feature #9258: Create role by copy
* Feature #9419: Group/sort the issue list by user/version-format custom fields
* Feature #10362: Show images in repositories inline when clicking the 'View' link

test/object_helpers.rb
module ObjectHelpers
  def User.generate!(attributes={})
    @generated_user_login ||= 'user0'
    @generated_user_login.succ!
    user = User.new(attributes)
    user.login = @generated_user_login if user.login.blank?
    user.mail = "#{@generated_user_login}@example.com" if user.mail.blank?
    user.firstname = "Bob" if user.firstname.blank?
    user.lastname = "Doe" if user.lastname.blank?
    yield user if block_given?
    user.save!
    user
  end

  def User.add_to_project(user, project, roles=nil)
    roles = Role.find(1) if roles.nil?
    roles = [roles] unless roles.is_a?(Array)
    Member.create!(:principal => user, :project => project, :roles => roles)
  end

  def Group.generate!(attributes={})
    @generated_group_name ||= 'Group 0'
    @generated_group_name.succ!
    group = Group.new(attributes)
    group.name = @generated_group_name if group.name.blank?
    yield group if block_given?
    group.save!
    group
  end

  def Project.generate!(attributes={})
    @generated_project_identifier ||= 'project-0000'
    @generated_project_identifier.succ!
    project = Project.new(attributes)
    project.name = @generated_project_identifier if project.name.blank?
    project.identifier = @generated_project_identifier if project.identifier.blank?
    yield project if block_given?
    project.save!
    project
  end

  def Tracker.generate!(attributes={})
    @generated_tracker_name ||= 'Tracker 0'
    @generated_tracker_name.succ!
    tracker = Tracker.new(attributes)
    tracker.name = @generated_tracker_name if tracker.name.blank?
    yield tracker if block_given?
    tracker.save!
    tracker
  end

  def Role.generate!(attributes={})
    @generated_role_name ||= 'Role 0'
    @generated_role_name.succ!
    role = Role.new(attributes)
    role.name = @generated_role_name if role.name.blank?
    yield role if block_given?
    role.save!
    role
  end

  def Issue.generate!(attributes={})
    issue = Issue.new(attributes)
    issue.project ||= Project.find(1)
    issue.tracker ||= issue.project.trackers.first
    issue.subject = 'Generated' if issue.subject.blank?
    issue.author ||= User.find(2)
    yield issue if block_given?
    issue.save!
    issue
  end

  # Generate an issue for a project, using its trackers
  def Issue.generate_for_project!(project, attributes={})
    issue = Issue.new(attributes) do |issue|
      issue.project = project
      issue.tracker = project.trackers.first unless project.trackers.empty?
      issue.subject = 'Generated' if issue.subject.blank?
      issue.author ||= User.find(2)
      yield issue if block_given?
    end
    issue.save!
    issue
  end

  # Generates an issue with some children and a grandchild
  def Issue.generate_with_descendants!(project, attributes={})
    issue = Issue.generate_for_project!(project, attributes)
    child = Issue.generate_for_project!(project, :subject => 'Child1', :parent_issue_id => issue.id)
    Issue.generate_for_project!(project, :subject => 'Child2', :parent_issue_id => issue.id)
    Issue.generate_for_project!(project, :subject => 'Child11', :parent_issue_id => child.id)
    issue.reload
  end

  def Version.generate!(attributes={})
    @generated_version_name ||= 'Version 0'
    @generated_version_name.succ!
    version = Version.new(attributes)
    version.name = @generated_version_name if version.name.blank?
    yield version if block_given?
    version.save!
    version
  end

  def AuthSource.generate!(attributes={})
    @generated_auth_source_name ||= 'Auth 0'
    @generated_auth_source_name.succ!
    source = AuthSource.new(attributes)
    source.name = @generated_auth_source_name if source.name.blank?
    yield source if block_given?
    source.save!
    source
  end

  def Board.generate!(attributes={})
    @generated_board_name ||= 'Forum 0'
    @generated_board_name.succ!
    board = Board.new(attributes)
    board.name = @generated_board_name if board.name.blank?
    board.description = @generated_board_name if board.description.blank?
    yield board if block_given?
    board.save!
    board
  end
end


lib/redmine/version.rb
require 'rexml/document'

module Redmine
  module VERSION #:nodoc:
    MAJOR = 2
    MINOR = 1
    TINY  = 3

    # Branch values:
    # * official release: nil
    # * stable branch:    stable
    # * trunk:            devel
    BRANCH = 'stable'

    def self.revision
      revision = nil
      entries_path = "#{Rails.root}/.svn/entries"
      if File.readable?(entries_path)
        begin
          f = File.open(entries_path, 'r')
          entries = f.read
          f.close
          if entries.match(%r{^\d+})
            revision = $1.to_i if entries.match(%r{^\d+\s+dir\s+(\d+)\s})
          else
            xml = REXML::Document.new(entries)
            revision =
              xml.elements['wc-entries'].elements[1].attributes['revision'].to_i
          end
        rescue
          # Could not find the current revision
        end
      end
      revision
    end

    REVISION = self.revision
    ARRAY    = [MAJOR, MINOR, TINY, BRANCH, REVISION].compact
    STRING   = ARRAY.join('.')

    def self.to_a; ARRAY  end
    def self.to_s; STRING end
  end
end


app/models/principal.rb
# Redmine - project management software
# Copyright (C) 2006-2012  Jean-Philippe Lang
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

class Principal < ActiveRecord::Base
  self.table_name = "#{table_name_prefix}users#{table_name_suffix}"

  has_many :members, :foreign_key => 'user_id', :dependent => :destroy
  has_many :memberships, :class_name => 'Member', :foreign_key => 'user_id', :include => [ :project, :roles ], :conditions => "#{Project.table_name}.status<>#{Project::STATUS_ARCHIVED}", :order => "#{Project.table_name}.name"
  has_many :projects, :through => :memberships
  has_many :issue_categories, :foreign_key => 'assigned_to_id', :dependent => :nullify

  # Groups and active users
  scope :active, :conditions => "#{Principal.table_name}.status = 1"

  scope :like, lambda {|q|
    if q.blank?
      {}
    else
      q = q.to_s
      pattern = "%#{q}%"
      sql = "LOWER(login) LIKE LOWER(:p) OR LOWER(firstname) LIKE LOWER(:p) OR LOWER(lastname) LIKE LOWER(:p) OR LOWER(mail) LIKE LOWER(:p)"
      params = {:p => pattern}
      if q =~ /^(.+)\s+(.+)$/
        a, b = "#{$1}%", "#{$2}%"
        sql << " OR (LOWER(firstname) LIKE LOWER(:a) AND LOWER(lastname) LIKE LOWER(:b)) OR (LOWER(firstname) LIKE LOWER(:b) AND LOWER(lastname) LIKE LOWER(:a))"
        params.merge!(:a => a, :b => b)
      end
      {:conditions => [sql, params]}
    end
  }

  # Principals that are members of a collection of projects
  scope :member_of, lambda {|projects|
    projects = [projects] unless projects.is_a?(Array)
    if projects.empty?
      {:conditions => "1=0"}
    else
      ids = projects.map(&:id)
      {:conditions => ["#{Principal.table_name}.status = 1 AND #{Principal.table_name}.id IN (SELECT DISTINCT user_id FROM #{Member.table_name} WHERE project_id IN (?))", ids]}
    end
  }
  # Principals that are not members of projects
  scope :not_member_of, lambda {|projects|
    projects = [projects] unless projects.is_a?(Array)
    if projects.empty?
      {:conditions => "1=0"}
    else
      ids = projects.map(&:id)
      {:conditions => ["#{Principal.table_name}.id NOT IN (SELECT DISTINCT user_id FROM #{Member.table_name} WHERE project_id IN (?))", ids]}
    end
  }

  before_create :set_default_empty_values

  def name(formatter = nil)
    to_s
  end

  def <=>(principal)
    if principal.nil?
      -1
    elsif self.class.name == principal.class.name
      self.to_s.downcase <=> principal.to_s.downcase
    else
      # groups after users
      principal.class.name <=> self.class.name
    end
  end

  protected

  # Make sure we don't try to insert NULL values (see #4632)
  def set_default_empty_values
    self.login ||= ''
    self.hashed_password ||= ''
    self.firstname ||= ''
    self.lastname ||= ''
    self.mail ||= ''
    true
  end
end


config/routes.rb
# Redmine - project management software
# Copyright (C) 2006-2012  Jean-Philippe Lang
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

RedmineApp::Application.routes.draw do
  root :to => 'welcome#index', :as => 'home'

  match 'login', :to => 'account#login', :as => 'signin'
  match 'logout', :to => 'account#logout', :as => 'signout'
  match 'account/register', :to => 'account#register', :via => [:get, :post], :as => 'register'
  match 'account/lost_password', :to => 'account#lost_password', :via => [:get, :post], :as => 'lost_password'
  match 'account/activate', :to => 'account#activate', :via => :get

  match '/news/preview', :controller => 'previews', :action => 'news', :as => 'preview_news'
  match '/issues/preview/new/:project_id', :to => 'previews#issue', :as => 'preview_new_issue'
  match '/issues/preview/edit/:id', :to => 'previews#issue', :as => 'preview_edit_issue'
  match '/issues/preview', :to => 'previews#issue', :as => 'preview_issue'

  match 'projects/:id/wiki', :to => 'wikis#edit', :via => :post
  match 'projects/:id/wiki/destroy', :to => 'wikis#destroy', :via => [:get, :post]

  match 'boards/:board_id/topics/new', :to => 'messages#new', :via => [:get, :post]
  get 'boards/:board_id/topics/:id', :to => 'messages#show', :as => 'board_message'
  match 'boards/:board_id/topics/quote/:id', :to => 'messages#quote', :via => [:get, :post]
  get 'boards/:board_id/topics/:id/edit', :to => 'messages#edit'

  post 'boards/:board_id/topics/preview', :to => 'messages#preview'
  post 'boards/:board_id/topics/:id/replies', :to => 'messages#reply'
  post 'boards/:board_id/topics/:id/edit', :to => 'messages#edit'
  post 'boards/:board_id/topics/:id/destroy', :to => 'messages#destroy'

  # Misc issue routes. TODO: move into resources
  match '/issues/auto_complete', :to => 'auto_completes#issues', :via => :get, :as => 'auto_complete_issues'
  match '/issues/context_menu', :to => 'context_menus#issues', :as => 'issues_context_menu'
  match '/issues/changes', :to => 'journals#index', :as => 'issue_changes'
  match '/issues/:id/quoted', :to => 'journals#new', :id => /\d+/, :via => :post, :as => 'quoted_issue'

  match '/journals/diff/:id', :to => 'journals#diff', :id => /\d+/, :via => :get
  match '/journals/edit/:id', :to => 'journals#edit', :id => /\d+/, :via => [:get, :post]

  match '/projects/:project_id/issues/gantt', :to => 'gantts#show'
  match '/issues/gantt', :to => 'gantts#show'

  match '/projects/:project_id/issues/calendar', :to => 'calendars#show'
  match '/issues/calendar', :to => 'calendars#show'

  match 'projects/:id/issues/report', :to => 'reports#issue_report', :via => :get
  match 'projects/:id/issues/report/:detail', :to => 'reports#issue_report_details', :via => :get

  match 'my/account', :controller => 'my', :action => 'account', :via => [:get, :post]
  match 'my/account/destroy', :controller => 'my', :action => 'destroy', :via => [:get, :post]
  match 'my/page', :controller => 'my', :action => 'page', :via => :get
  match 'my', :controller => 'my', :action => 'index', :via => :get # Redirects to my/page
  match 'my/reset_rss_key', :controller => 'my', :action => 'reset_rss_key', :via => :post
  match 'my/reset_api_key', :controller => 'my', :action => 'reset_api_key', :via => :post
  match 'my/password', :controller => 'my', :action => 'password', :via => [:get, :post]
  match 'my/page_layout', :controller => 'my', :action => 'page_layout', :via => :get
  match 'my/add_block', :controller => 'my', :action => 'add_block', :via => :post
  match 'my/remove_block', :controller => 'my', :action => 'remove_block', :via => :post
  match 'my/order_blocks', :controller => 'my', :action => 'order_blocks', :via => :post

  resources :users
  match 'users/:id/memberships/:membership_id', :to => 'users#edit_membership', :via => :put, :as => 'user_membership'
  match 'users/:id/memberships/:membership_id', :to => 'users#destroy_membership', :via => :delete
  match 'users/:id/memberships', :to => 'users#edit_membership', :via => :post, :as => 'user_memberships'

  match 'watchers/new', :controller=> 'watchers', :action => 'new', :via => :get
  match 'watchers', :controller=> 'watchers', :action => 'create', :via => :post
  match 'watchers/append', :controller=> 'watchers', :action => 'append', :via => :post
  match 'watchers/destroy', :controller=> 'watchers', :action => 'destroy', :via => :post
  match 'watchers/watch', :controller=> 'watchers', :action => 'watch', :via => :post
  match 'watchers/unwatch', :controller=> 'watchers', :action => 'unwatch', :via => :post
  match 'watchers/autocomplete_for_user', :controller=> 'watchers', :action => 'autocomplete_for_user', :via => :get

  match 'projects/:id/settings/:tab', :to => "projects#settings"

  resources :projects do
    member do
      get 'settings'
      post 'modules'
      post 'archive'
      post 'unarchive'
      post 'close'
      post 'reopen'
      match 'copy', :via => [:get, :post]
    end

    resources :memberships, :shallow => true, :controller => 'members', :only => [:index, :show, :create, :update, :destroy] do
      collection do
        get 'autocomplete'
      end
    end

    resource :enumerations, :controller => 'project_enumerations', :only => [:update, :destroy]

    match 'issues/:copy_from/copy', :to => 'issues#new'
    resources :issues, :only => [:index, :new, :create] do
      resources :time_entries, :controller => 'timelog' do
        collection do
          get 'report'
        end
      end
    end
    # issue form update
    match 'issues/new', :controller => 'issues', :action => 'new', :via => [:put, :post], :as => 'issue_form'

    resources :files, :only => [:index, :new, :create]

    resources :versions, :except => [:index, :show, :edit, :update, :destroy] do
      collection do
        put 'close_completed'
      end
    end
    match 'versions.:format', :to => 'versions#index'
    match 'roadmap', :to => 'versions#index', :format => false
    match 'versions', :to => 'versions#index'

    resources :news, :except => [:show, :edit, :update, :destroy]
    resources :time_entries, :controller => 'timelog' do
      get 'report', :on => :collection
    end
    resources :queries, :only => [:new, :create]
    resources :issue_categories, :shallow => true
    resources :documents, :except => [:show, :edit, :update, :destroy]
    resources :boards
    resources :repositories, :shallow => true, :except => [:index, :show] do
      member do
        match 'committers', :via => [:get, :post]
      end
    end

    match 'wiki/index', :controller => 'wiki', :action => 'index', :via => :get
    match 'wiki/:id/diff/:version/vs/:version_from', :controller => 'wiki', :action => 'diff'
    match 'wiki/:id/diff/:version', :controller => 'wiki', :action => 'diff'
    resources :wiki, :except => [:index, :new, :create] do
      member do
        get 'rename'
        post 'rename'
        get 'history'
        get 'diff'
        match 'preview', :via => [:post, :put]
        post 'protect'
        post 'add_attachment'
      end
      collection do
        get 'export'
        get 'date_index'
      end
    end
    match 'wiki', :controller => 'wiki', :action => 'show', :via => :get
    match 'wiki/:id/annotate/:version', :controller => 'wiki', :action => 'annotate'
  end

  resources :issues do
    collection do
      match 'bulk_edit', :via => [:get, :post]
      post 'bulk_update'
    end
    resources :time_entries, :controller => 'timelog' do
      collection do
        get 'report'
      end
    end
    resources :relations, :shallow => true, :controller => 'issue_relations', :only => [:index, :show, :create, :destroy]
  end
  match '/issues', :controller => 'issues', :action => 'destroy', :via => :delete

  resources :queries, :except => [:show]

  resources :news, :only => [:index, :show, :edit, :update, :destroy]
  match '/news/:id/comments', :to => 'comments#create', :via => :post
  match '/news/:id/comments/:comment_id', :to => 'comments#destroy', :via => :delete

  resources :versions, :only => [:show, :edit, :update, :destroy] do
    post 'status_by', :on => :member
  end

  resources :documents, :only => [:show, :edit, :update, :destroy] do
    post 'add_attachment', :on => :member
  end

  match '/time_entries/context_menu', :to => 'context_menus#time_entries', :as => :time_entries_context_menu

  resources :time_entries, :controller => 'timelog', :except => :destroy do
    collection do
      get 'report'
      get 'bulk_edit'
      post 'bulk_update'
    end
  end
  match '/time_entries/:id', :to => 'timelog#destroy', :via => :delete, :id => /\d+/
  # TODO: delete /time_entries for bulk deletion
  match '/time_entries/destroy', :to => 'timelog#destroy', :via => :delete

  # TODO: port to be part of the resources route(s)
  match 'projects/:id/settings/:tab', :to => 'projects#settings', :via => :get

  get 'projects/:id/activity', :to => 'activities#index'
  get 'projects/:id/activity.:format', :to => 'activities#index'
  get 'activity', :to => 'activities#index'

  # repositories routes
  get 'projects/:id/repository/:repository_id/statistics', :to => 'repositories#stats'
  get 'projects/:id/repository/:repository_id/graph', :to => 'repositories#graph'

  get 'projects/:id/repository/:repository_id/changes(/*path(.:ext))',
      :to => 'repositories#changes'

  get 'projects/:id/repository/:repository_id/revisions/:rev', :to => 'repositories#revision'
  get 'projects/:id/repository/:repository_id/revision', :to => 'repositories#revision'
  post   'projects/:id/repository/:repository_id/revisions/:rev/issues', :to => 'repositories#add_related_issue'
  delete 'projects/:id/repository/:repository_id/revisions/:rev/issues/:issue_id', :to => 'repositories#remove_related_issue'
  get 'projects/:id/repository/:repository_id/revisions', :to => 'repositories#revisions'
  get 'projects/:id/repository/:repository_id/revisions/:rev/:action(/*path(.:ext))',
      :controller => 'repositories',
      :format => false,
      :constraints => {
            :action => /(browse|show|entry|raw|annotate|diff)/,
            :rev    => /[a-z0-9\.\-_]+/
          }

  get 'projects/:id/repository/statistics', :to => 'repositories#stats'
  get 'projects/:id/repository/graph', :to => 'repositories#graph'

  get 'projects/:id/repository/changes(/*path(.:ext))',
      :to => 'repositories#changes'

  get 'projects/:id/repository/revisions', :to => 'repositories#revisions'
  get 'projects/:id/repository/revisions/:rev', :to => 'repositories#revision'
  get 'projects/:id/repository/revision', :to => 'repositories#revision'
  post   'projects/:id/repository/revisions/:rev/issues', :to => 'repositories#add_related_issue'
  delete 'projects/:id/repository/revisions/:rev/issues/:issue_id', :to => 'repositories#remove_related_issue'
  get 'projects/:id/repository/revisions/:rev/:action(/*path(.:ext))',
      :controller => 'repositories',
      :format => false,
      :constraints => {
            :action => /(browse|show|entry|raw|annotate|diff)/,
            :rev    => /[a-z0-9\.\-_]+/
          }
  get 'projects/:id/repository/:repository_id/:action(/*path(.:ext))',
      :controller => 'repositories',
      :action => /(browse|show|entry|raw|changes|annotate|diff)/
  get 'projects/:id/repository/:action(/*path(.:ext))',
      :controller => 'repositories',
      :action => /(browse|show|entry|raw|changes|annotate|diff)/

  get 'projects/:id/repository/:repository_id', :to => 'repositories#show', :path => nil
  get 'projects/:id/repository', :to => 'repositories#show', :path => nil

  # additional routes for having the file name at the end of url
  match 'attachments/:id/:filename', :controller => 'attachments', :action => 'show', :id => /\d+/, :filename => /.*/, :via => :get
  match 'attachments/download/:id/:filename', :controller => 'attachments', :action => 'download', :id => /\d+/, :filename => /.*/, :via => :get
  match 'attachments/download/:id', :controller => 'attachments', :action => 'download', :id => /\d+/, :via => :get
  match 'attachments/thumbnail/:id(/:size)', :controller => 'attachments', :action => 'thumbnail', :id => /\d+/, :via => :get, :size => /\d+/
  resources :attachments, :only => [:show, :destroy]

  resources :groups do
    member do
      get 'autocomplete_for_user'
    end
  end

  match 'groups/:id/users', :controller => 'groups', :action => 'add_users', :id => /\d+/, :via => :post, :as => 'group_users'
  match 'groups/:id/users/:user_id', :controller => 'groups', :action => 'remove_user', :id => /\d+/, :via => :delete, :as => 'group_user'
  match 'groups/destroy_membership/:id', :controller => 'groups', :action => 'destroy_membership', :id => /\d+/, :via => :post
  match 'groups/edit_membership/:id', :controller => 'groups', :action => 'edit_membership', :id => /\d+/, :via => :post

  resources :trackers, :except => :show do
    collection do
      match 'fields', :via => [:get, :post]
    end
  end
  resources :issue_statuses, :except => :show do
    collection do
      post 'update_issue_done_ratio'
    end
  end
  resources :custom_fields, :except => :show
  resources :roles, :except => :show do
    collection do
      match 'permissions', :via => [:get, :post]
    end
  end
  resources :enumerations, :except => :show

  get 'projects/:id/search', :controller => 'search', :action => 'index'
  get 'search', :controller => 'search', :action => 'index'

  match 'mail_handler', :controller => 'mail_handler', :action => 'index', :via => :post

  match 'admin', :controller => 'admin', :action => 'index', :via => :get
  match 'admin/projects', :controller => 'admin', :action => 'projects', :via => :get
  match 'admin/plugins', :controller => 'admin', :action => 'plugins', :via => :get
  match 'admin/info', :controller => 'admin', :action => 'info', :via => :get
  match 'admin/test_email', :controller => 'admin', :action => 'test_email', :via => :get
  match 'admin/default_configuration', :controller => 'admin', :action => 'default_configuration', :via => :post

  resources :auth_sources do
    member do
      get 'test_connection'
    end
  end

  match 'workflows', :controller => 'workflows', :action => 'index', :via => :get
  match 'workflows/edit', :controller => 'workflows', :action => 'edit', :via => [:get, :post]
  match 'workflows/permissions', :controller => 'workflows', :action => 'permissions', :via => [:get, :post]
  match 'workflows/copy', :controller => 'workflows', :action => 'copy', :via => [:get, :post]
  match 'settings', :controller => 'settings', :action => 'index', :via => :get
  match 'settings/edit', :controller => 'settings', :action => 'edit', :via => [:get, :post]
  match 'settings/plugin/:id', :controller => 'settings', :action => 'plugin', :via => [:get, :post]

  match 'sys/projects', :to => 'sys#projects', :via => :get
  match 'sys/projects/:id/repository', :to => 'sys#create_project_repository', :via => :post
  match 'sys/fetch_changesets', :to => 'sys#fetch_changesets', :via => :get

  match 'uploads', :to => 'attachments#upload', :via => :post

  get 'robots.txt', :to => 'welcome#robots'

  Dir.glob File.expand_path("plugins/*", Rails.root) do |plugin_dir|
    file = File.join(plugin_dir, "config/routes.rb")
    if File.exists?(file)
      begin
        instance_eval File.read(file)
      rescue Exception => e
        puts "An error occurred while loading the routes definition of #{File.basename(plugin_dir)} plugin (#{file}): #{e.message}."
        exit 1
      end
    end
  end
end
